### PR TITLE
refactor(precompile): dynamically generate precompile abi <> go implementation mapping

### PR DESCRIPTION
**Pull request by shampoobera**
just how geth uses reflection to register the different APIs and JSON-RPC, we do the same but for precompiles

### CODE REVIEW

**Comment by itsdevbear** [comment]
@shampoobera lgtm

**Review by itsdevbear — APPROVED** [review]
lgtm good shit

**Comment by itsdevbear** [comment]
Make sure to fully remove RequiredGas tho. I think we should just deprecate it for simplicity.

**Comment by itsdevbear** [comment]
or alternatively, allow precompiles to specific intrinsic gas or something, meaning it gets consumed the same for all functions no matter what?

File: eth/core/precompile/method_test.go [hunk 1]
```diff
@@ -52,17 +51,6 @@ var _ = Describe("Method", func() {
 			err := methodMissingFunc.ValidateBasic()
 			Expect(err).To(HaveOccurred())
 		})
-
-		It("should error on given abi method", func() {
```
Comment (itsdevbear) [inline eth/core/precompile/method_test.go:56]: @shampoobera I think removing this test is bad, if there are unused abi methods, they it should error.

**Review by itsdevbear — DISMISSED** [review]
abi safety thing

**Comment by calbera** [comment]
> Make sure to fully remove RequiredGas tho. I think we should just deprecate it for simplicity.

Only concern from deprecating this is “helper” functions that behave like the geth default precompiles. If they’re 0 gas it’s a way to DOS

File: cosmos/precompile/staking/helpers.go
Comment (calbera) [inline cosmos/precompile/staking/helpers.go:1]: revert all changes to this file, handled all the cases in a diff PR

Comment (itsdevbear) [reply cosmos/precompile/staking/helpers.go:1]: no don't, that change is already in main, @shampoobera move target to `main`.

File: eth/core/precompile/method_test.go (hunk 1 above)
Comment (itsdevbear) [reply eth/core/precompile/method_test.go:56]: oh I mean the test is bad now, but we shouldnt remove the test you feel (just fix it to actually use a bad one)

i'd also add a test for if there are unused abi methods in the solidity file (a.k.a, not matched to something in the go precompile` it should also error.

Comment (shampoobera) [reply eth/core/precompile/method_test.go:56]: we already do that in `factories_test.go`

https://github.com/berachain/polaris/pull/869/files#diff-623c4a231b9e56ca6de325b03c827a6c7218c9ee347a7a518acba6861ab6df7cR79

**Comment by itsdevbear** [comment]
i think we wanna do a "base gas" or something that gets consumed for all functions or something

File: .vscode/launch.json
Comment (calbera) [inline .vscode/launch.json:1]: what is this

File: cosmos/testing/integration/precompile/bank/bank_test.go [hunk 2]
```diff
@@ -98,8 +98,10 @@ var _ = Describe("Bank", func() {
 		Expect(balance.Cmp(big.NewInt(1000000000000000000))).To(Equal(0))
 
 		// Send 1000 bera from alice to charlie
+		txr := tf.GenerateTransactOpts("alice")
+		// txr.GasLimit = 10000000
 		_, err = bankPrecompile.Send(
-			tf.GenerateTransactOpts("alice"),
+			txr,
```
Comment (calbera) [inline cosmos/testing/integration/precompile/bank/bank_test.go:104]: undo these

File: eth/core/precompile/factories.go [hunk 3]
```diff
@@ -150,3 +154,109 @@ func (sf *StatefulFactory) buildIdsToMethods(
 	}
 	return idsToMethods, nil
 }
+
+// GeneratePrecompileMethods generates the methods for the given Precompile's ABI.
+func GeneratePrecompileMethods(abiMethods map[string]abi.Method, contractImpl reflect.Value) (Methods, error) {
+	return suitableMethods(abiMethods, contractImpl)
+}
```
Comment (calbera) [inline eth/core/precompile/factories.go:161]: don't need this, just call `suitable` directly

File: eth/core/precompile/polar_context.go
Comment (calbera) [inline eth/core/precompile/polar_context.go:1]: don't need this rn

File: eth/core/precompile/factories.go [hunk 4]
```diff
@@ -150,3 +154,109 @@ func (sf *StatefulFactory) buildIdsToMethods(
...
+func isBaseContractMethodOrUnexported(implMethod reflect.Method) bool {
+	return implMethod.Name == "RegistryKey" ||
+		implMethod.Name == "ABIMethods" ||
+		implMethod.Name == "ABIEvents" ||
+		implMethod.Name == "CustomValueDecoders" ||
+		implMethod.Name == "PrecompileMethods" ||
+		implMethod.Name == "SetPlugin" ||
+		implMethod.Name == "GetPlugin" ||
+		implMethod.PkgPath != ""
+}
```
Comment (calbera) [inline eth/core/precompile/factories.go:211]: don't need this

File: eth/core/precompile/factories.go [hunk 5]
```diff
@@ -150,3 +154,109 @@ func (sf *StatefulFactory) buildIdsToMethods(
...
+func suitableMethods(pcABI map[string]abi.Method, contractImpl reflect.Value) (Methods, error) {
+	contractImplType := contractImpl.Type()
+	var methods Methods
+	for m := 0; m < contractImplType.NumMethod(); m++ { // iterate through all of the impl's methods
+		implMethod := contractImplType.Method(m) // grab the Impl's current method
+
+		if isBaseContractMethodOrUnexported(implMethod) {
+			continue // ignore BaseContract's methods, they clearly won't be in the abi
+			// also ignore methods that are not exported
+		}
```
Comment (calbera) [inline eth/core/precompile/factories.go:177]: useless check

File: eth/core/precompile/factories.go [hunk 6]
```diff
@@ -150,3 +154,109 @@ func (sf *StatefulFactory) buildIdsToMethods(
...
+			}
+			toExecute := newExecute(implMethod) // grab the actual function
+			methods = append(methods, &Method{
+				AbiMethod: &abiMethod,
+				AbiSig:    abiMethod.Sig,
+				Execute:   toExecute,
+			}) // add it to the list of methods
+		} else {
+			return nil, errors.Wrap(ErrNoPrecompileMethodForABIMethod, implMethodName)
+		}
```
Comment (calbera) [inline eth/core/precompile/factories.go:192]: just check at the end that the size of `methods` is equal to size of abi methods

File: eth/core/precompile/factories.go [hunk 7]
```diff
@@ -150,3 +154,109 @@ func (sf *StatefulFactory) buildIdsToMethods(
...
+	// // uahsdiouashduiosahd
+	// return errors.Wrap(ErrNoPrecompileMethodForABIMethod, abiMethod.Sig)
+	// }
+	// }
+	return nil
+}
+
+func newExecute(fn reflect.Method) reflect.Value {
+	return fn.Func
+}
```
Comment (calbera) [inline eth/core/precompile/factories.go:246]: dont need in this PR

File: eth/core/precompile/method.go [hunk 8]
```diff
@@ -80,13 +74,64 @@ type Method struct {
 // ValidateBasic returns an error if this a precompile `Method` has invalid fields.
 func (m *Method) ValidateBasic() error {
 	// ensure all required fields are nonempty
-	if len(m.AbiSig) == 0 || m.AbiMethod != nil || m.Execute == nil {
+	if len(m.AbiSig) == 0 || !m.Execute.IsValid() || m.Execute.IsNil() {
 		return ErrIncompleteMethod
 	}
 
 	return nil
 }
```
Comment (calbera) [inline eth/core/precompile/method.go:82]: unneeded

File: eth/core/precompile/method.go [hunk 9]
```diff
@@ -80,13 +74,64 @@ type Method struct {
...
+
+	// If the precompile returns something wrong (e.g. wrong number of results), the error is
+	// returned to the caller.
+	if len(results) != 2 { //nolint:gomnd // its okay.
+		return nil, errors.Wrapf(
+			vm.ErrExecutionReverted,
+			"vm error: precompile [%s] returned [%d] results, expected 2",
+			debug.GetFnName(m.Execute.Interface()), len(results),
+		)
+	}
```
Comment (calbera) [inline eth/core/precompile/method.go:109]: this should not happen here. at registration you should statically verify that there are only 2 return values on that function

Comment (shampoobera) [reply eth/core/precompile/method.go:99]: results is after the actual call is done, this doesn't have to do with the return values of the function. the two is the first return value and the second is possible errors just the way that reflect works

Comment (itsdevbear) [reply eth/core/precompile/method.go:99]: yeah @calbera I think this is more a reflect thing, although @shampoobera idk if we need it, may be a redudant check (although always better to check more than less)

File: .vscode/launch.json
Comment (calbera) [inline .vscode/launch.json:1]: remove changes

File: eth/core/precompile/factories.go [hunk 10]
```diff
@@ -92,61 +97,99 @@ func (sf *StatefulFactory) Build(
...
-		if i == len(precompileMethods) {
-			return nil, errors.Wrap(ErrNoPrecompileMethodForABIMethod, abiMethod.Sig)
-		}
+	}
 
-		// attach the ABI method to the precompile method for stateful container to handle
-		precompileMethod.AbiMethod = &abiMethod
-		idsToMethods[utils.UnsafeBytesToStr(abiMethod.ID)] = precompileMethod
+	for _, abiMethod := range pcABI { // iterate through all of the ABI's methods
+		if _, found := idsToMethods[utils.UnsafeBytesToStr(abiMethod.ID)]; !found { // if the method is not found in the ABI
```
Comment (calbera) [inline eth/core/precompile/factories.go:148]: style nitpick: add these type of comments on top of/before the code block

File: eth/core/precompile/factories.go [hunk 11]
```diff
@@ -92,61 +97,99 @@ func (sf *StatefulFactory) Build(
...
-			return nil, err
-		}
-	}
-
-	// match every ABI method to corresponding precompile method
+// This function matches each Go implementation of the Precompile
+// to the ABI's respective function.
+// It first searches for the ABI function in the Go implementation. If no find, then panic.
+// It then performs some basic validation on the implemented function
+// Then, the implemented function's arguments are checked against the ABI's arguments' types.
```
Comment (calbera) [inline eth/core/precompile/factories.go:120]: style nitpick: write the godoc comments as a paragraph. so extend each of the lines to 100 characters

File: eth/core/precompile/factories.go [hunk 12]
```diff
@@ -92,61 +97,99 @@ func (sf *StatefulFactory) Build(
...
-				precompileMethod = precompileMethods[i]
-				break
+	for m := 0; m < contractImplType.NumMethod(); m++ { // iterate through all of the impl's methods
+		implMethod := contractImplType.Method(m) // grab the Impl's current method
+
+		// write me some logic that checks if the second return value of the implmethod is an error.
+		// if it is, then return an error. if it isn't, then return nil.
+		// this is because the ABI methods are not allowed to return an error.
+		// if the impl method returns an error, then it is not a valid precompile method.
+		// if the impl method does not return an error, then it is a valid precompile method.
```
Comment (calbera) [inline eth/core/precompile/factories.go:131]: remove

File: eth/core/precompile/method.go [hunk 13]
```diff
@@ -69,22 +63,48 @@ type Method struct {
 
 	// Execute is the precompile's executable which will execute the logic of the implemented
 	// ABI method.
-	Execute Executable
+	Execute reflect.Value
 
 	// RequiredGas is the amount of gas (as a `uint64`) used up by the execution of `Execute`.
 	// This field is optional; if left empty, the precompile's executable should consume gas using
 	// the native gas meter.
 	RequiredGas uint64
```
Comment (calbera) [inline eth/core/precompile/method.go:71]: remove

File: eth/core/precompile/method.go [hunk 14]
```diff
@@ -40,19 +41,12 @@ import (
...
-	evm EVM,
-	caller common.Address,
-	value *big.Int,
-	readonly bool,
-	args ...any,
-) (ret []any, err error)
-
 // Method is a struct that contains the required information for the EVM to execute a stateful
 // precompiled contract method.
 type Method struct {
```
Comment (calbera) [inline eth/core/precompile/method.go:52]: make all the fields of Method private

File: eth/core/precompile/factories.go [hunk 15]
```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
...
+	}
+
+	return string(ret)
+}
+
+// checkReturnTypes checks if the precompile method returns a []any and an error.
+// If it does not, then an error is returned.
+func checkReturnTypes(implMethod reflect.Method) error {
+	if implMethod.Type.NumOut() != 2 { //nolint:gomnd // it's okay.
+		return errors.New("precompile methods must return ([]any, error), but found wrong number of return types for precompile method: " + //nolint:lll // it's okay.
```
Comment (calbera) [inline eth/core/precompile/factories.go:167]: fmt.Errorf

File: eth/core/precompile/factories.go [hunk 16]
```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
...
+			implMethod.Name)
+	}
+	firstReturnType := implMethod.Type.Out(0)
+	secondReturnType := implMethod.Type.Out(1)
+
+	if firstReturnType.Kind() != reflect.Slice { // check if the first return type is a []any
+		return errors.New("first parameter should be []any, but found " +
```
Comment (calbera) [inline eth/core/precompile/factories.go:174]: fmt.Errorf

File: eth/core/precompile/factories.go [hunk 17]
```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
...
+			firstReturnType.String() + " for precompile method: " + implMethod.Name)
+	} else if firstReturnType.Elem().Kind() != reflect.Interface { // if it is but it is not an any...
+		return errors.New("first parameter should be []any, but found " +
```
Comment (calbera) [inline eth/core/precompile/factories.go:177]: fmt.Errorf

File: eth/core/precompile/factories.go [hunk 18]
```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
...
+			firstReturnType.String() + " for precompile method: " + implMethod.Name)
+	}
+
+	if secondReturnType.Name() != "error" { // if the second return value is not an error
+		return errors.New("second parameter should be error, but found " +
```
Comment (calbera) [inline eth/core/precompile/factories.go:182]: fmt.Errorf

File: eth/core/precompile/factories.go [hunk 19]
```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
...
+
 	return idsToMethods, nil
 }
+
+// formatName converts to first character of name to lowercase. If the first
+// three characters are "ERC" or "ABI" (which is p common), then it converts all
+// three to lowercase.
+func formatName(name string) string {
+	ret := []rune(name)
+	if name[:3] == "ERC" || name[:3] == "ABI" { // special case for ERC20, ERC721, etc.
```
Comment (calbera) [inline eth/core/precompile/factories.go:152]: this will panic if name is less than 3 chars long

File: eth/core/precompile/factories.go [hunk 20]
```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
...
+		if _, found := idsToMethods[utils.UnsafeBytesToStr(abiMethod.ID)]; !found {
+			return nil, errorslib.Wrap(ErrNoPrecompileMethodForABIMethod, abiMethod.Name)
+		}
 	}
+
 	return idsToMethods, nil
 }
+
+// formatName converts to first character of name to lowercase. If the first
+// three characters are "ERC" or "ABI" (which is p common), then it converts all
```
Comment (itsdevbear) [inline eth/core/precompile/factories.go:148]: this is probably a bad idea lmfaoooo, i can already see the gremlins this is gonna cause.

Comment (shampoobera) [reply eth/core/precompile/factories.go:148]: thoughts on making the go function name exactly match (and subsequently unexport them) the solidity funcion names then? this was more of an edge case for functions like `ERC20AddressForCoinDenom` in `erc20.go`

Comment (itsdevbear) [reply eth/core/precompile/factories.go:148]: i think just make it be weird tbh at least it's consistent

**Review by calbera — APPROVED** [review]

File: eth/core/precompile/stateful.go [hunk 21]
```diff
@@ -83,55 +81,23 @@ func (sc *stateful) Run(
...
-	}
-
-	// Extract the method ID from the input and load the method.
-	method, found := sc.idsToMethods[utils.UnsafeBytesToStr(input[:NumBytesMethodID])]
-	if !found {
-		return 0
-	}
-
-	return method.RequiredGas
+func (sc *stateful) RequiredGas(_ []byte) uint64 {
```
Comment (itsdevbear) [inline eth/core/precompile/stateful.go:101]: add TODO: remove in later PR.

**Review by itsdevbear — APPROVED** [review]

**Review by itsdevbear — APPROVED** [review]

**Pull request merged**